tools/codediff/run_command.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
+# SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# run_command.sh - run a single command and record environment info
+#
+set -e
+set -o pipefail
+usage() {
+echo "Usage: $0 [-h] [-n <run_name>] -o <output_directory> -- command to run and arguments"
+}
+while getopts "n:o:h" arg
+do
+case $arg in
+n)
+runname=$OPTARG
+;;
+o)
+testdir=$OPTARG
+;;
+h | *)
+usage
+exit 1
+;;
+esac
+done
+if [[ -z $testdir ]]
+then
+echo "The output directory must be specified with -o"
+usage
+exit 1
+fi
+# getopts stops parsing if it sees "--" but does not shift. We can detect that case and record command
+while [[ $# -gt 0 ]]
+do
+if [[ "$1" == "--" ]]
+then
+shift
+break
+fi
+shift
+done
+testcmd=$*
+if [[ -z "$testcmd" ]]
+then
+usage
+exit 1
+fi
+if [[ -f "$testdir/command" ]]
+then
+echo -n "Skipping since $testdir/command exists. "
+echo "To re-run, remove $testdir or specify another and try again."
+exit 1
+fi
+mkdir -p "$testdir"
+movecudafiles() {
+mkdir -p "$1/cuda" "$1/ptx"
+find . -maxdepth 1 -name '__tmp_kernel*.cu' -print0 | xargs -0 mv -t "$1/cuda"
+find . -maxdepth 1 -name '__tmp_kernel*.ptx' -print0 | xargs -0 mv -t "$1/ptx"
+}
+removecudafiles() {
+tmpdir="./.nvfuser_run_command_tmp"
+mkdir -p "$tmpdir"
+movecudafiles "$tmpdir"
+rm -rf "$tmpdir"
+}
+stdoutfile="$testdir/incomplete-stdout-$(date +%Y%m%d_%H%M%S).log"
+cleanup() {
+numcu=$(find . -maxdepth 1 -name '__tmp_kernel*.cu' | wc -l)
+numptx=$(find . -maxdepth 1 -name '__tmp_kernel*.ptx' | wc -l)
+if (( numcu + numptx > 0 ))
+then
+echo "Interrupted. Removing $numcu temporary .cu files and $numptx temporary .ptx files"
+removecudafiles
+fi
+logbase=$(basename "$stdoutfile")
+# strip incomplete- from base name
+completelogbase=${logbase#incomplete-}
+mv "$stdoutfile" "$testdir/$completelogbase" 2> /dev/null || true
+}
+trap "cleanup" EXIT
+# Allow command to fail, but record exit code
+set +e
+$testcmd | tee "$stdoutfile"
```

Comment:
Can we append a ` | head -c 1M` to this line? This won't affect the `tee`-ed stdout file, but it limits the output printed to screen. Ref: https://serverfault.com/questions/277948/easy-way-to-limit-file-size-stdout-on-a-shell-script-level

Our CI has a log size limit of 5 MB. Setting this to 1 MB will make the printed logs from two runs not exceeding 2 MB. There are some weird problems happening when the CI log limit is exceeded.

Reply:
Great idea. We could also just not tee at all. The stdout logs are still there and could be collected as an artifact.

Reply:
Let's not add the `| head -c 1M` thing for now. I think I misunderstood how it works. It will cause SIGPIPE error if the full output exceeds the head limit. https://stackoverflow.com/questions/19120263/why-exit-code-141-with-grep-q

Reply:
I added the `-q` (quiet) arg to both `compare_codegen.sh` and `run_command.sh`. By default output will be shown. With `-q` we just print `Running: the command line...`.